### PR TITLE
Refactoring count

### PR DESCRIPTION
**Pull request by ghar1821**
This description will be updated as the refactoring progresses.

The main idea of this refactoring attempt is to split the huge count script, and somewhat turn it a bit more object oriented. 

What has been done so far:

- New module responsible for processing the parameters/arguments.
- A parameter class which store the parameters for the "count" functionality and a whole host of code which are related to the parameters (checking them or creating something else out of them).
- Move all the modules for the count into its own directory.

### CODE REVIEW

**Comment by iosonofabio** [comment]
This is going well. `runner.py` has no obvious reason to exist: `count.py` is part of the API, thereffore we cannot easily delete it, moreover `htseq-count` is already an empty shell basically calling `count.py`, we don't need another layer of emptiness behind that screen.

I think the most important piece missing at the moment is to separate two chunks in the `do_` function: overlapping the features, and what to do once the list of overlapping features `fs` is set. Those would ideally be separate functions.

Don't worry too much about the module structure (`core`, etc.). I'll take a stab at it once the PR is a little more advanced

**Comment by ghar1821** [comment]
Yeah I'm not exactly happy with the `runner.py`. The content can easily be dumped into `param_builder`. I think I'll do that. 

I was gonna eventually remove `count.py` after migrating the content into `count_reads_in_feature`. This should affect the `htseq-count` command as we just need to change the script. 
For the main API, users just need to (in the future) create `CountParameters` then run `do_count_reads_in_features`.

When you said 2 chunks in `do_` function, do you mean this bit?
```
# prepare to run count in parallel
    args = in_param.get_args_for_count()

    # Count reads in parallel
    if nprocesses > 1:
        with multiprocessing.Pool(nprocesses) as pool:
            results = pool.starmap(count_reads_single_file, args)
        results.sort(key=operator.itemgetter('isam_file_idx'))
    else:
        results = list(itertools.starmap(count_reads_single_file, args))
```
Yeah they need to be split as `run_parallel` and `run_serial`.

**Comment by ghar1821** [comment]
Some documentations still need to be updated.

TLDR;

- New objects ReadStatistics and ReadIO. The first handles the feature alignment and counting while the latter handles the input/output (preparing BAM/SAM input/output files, etc.).
- Migrated the content count_reads_single_file into a module within a new sub-package `count_features`.
- Refactored the content of original count_reads_single_file.
- Refactored the content of new count.

Test cases are passing.

image
